docs/migrate-to-nativecurlclient.md

```diff
@@ -0,0 +1,255 @@
++$body = $this->client->getLastResponseBody();
++
++// if response is XML, create a SimpleXMLElement object
++if ($body !== '' && 0 === strpos($this->client->getLastResponseContentType(), 'application/xml')) {
++    $dataAsXML = new \SimpleXMLElement($body);
++} else if ($body !== '' && 0 === strpos($this->client->getLastResponseContentType(), 'application/json')) {
++    try {
++        $dataAsJson = json_decode($body, true, 512, \JSON_THROW_ON_ERROR);
++    } catch (\JsonException $e) {
++        throw new \Exception('Error decoding body as JSON: '.$e->getMessage());
++    }
++} else {
++    $dataAsString = $body;
++}
+```
+
+### post() to requestPost()
+
+This example shows how you can parse the response body of a POST request.
+
+```diff
+-// We dont know if we will get xml or a string
+-$dataAsXmlOrString = $this->client->post($path, $data);
++$this->client->requestPost($path, $data);
++// $body contains the raw http body of the response
++$body = $this->client->getLastResponseBody();
++
++// if response is XML, create a SimpleXMLElement object
++if ($body !== '' && 0 === strpos($this->client->getLastResponseContentType(), 'application/xml')) {
++    $dataAsXML = new \SimpleXMLElement($body);
++} else {
++    $dataAsString = $body;
++}
+```
+
+### put() to requestPut()
+
+This example shows how you can parse the response body of a PUT request.
+
+```diff
+-// We dont know if we will get xml or a string
+-$dataAsXmlOrString = $this->client->put($path, $data);
++$this->client->requestPut($path, $data);
++// $body contains the raw http body of the response
++$body = $this->client->getLastResponseBody();
++
++// if response is XML, create a SimpleXMLElement object
++if ($body !== '' && 0 === strpos($this->client->getLastResponseContentType(), 'application/xml')) {
++    $dataAsXML = new \SimpleXMLElement($body);
++} else {
++    $dataAsString = $body;
++}
+```
+
+### delete() to requestDelete()
+
+This example shows how you can parse the response body of a DELETE request.
+
+```diff
+-$dataAsString = $this->client->delete($path);
++$this->client->requestDelte($path);
++$dataAsString = $this->client->getLastResponseBody();
+```
+
+### setImpersonateUser() to startImpersonateUser()
+
+If you are using the [Redmine user impersonation](https://www.redmine.org/projects/redmine/wiki/Rest_api#User-Impersonation)
+you have to change your code.
+
+```diff
+// impersonate the user `robin`
+-$client->setImpersonateUser('robin');
++$client->startImpersonateUser('robin');
+
+$userData = $client->getApi('user')->getCurrentUser();
+
+// Now stop impersonation
+-$client->setImpersonateUser(null);
++$client->stopImpersonateUser();
+```
+
+### setCheckSslCertificate() to setCurlOption()
+
+```diff
+-$client->setCheckSslCertificate(true);
++$client->setCurlOption(CURLOPT_SSL_VERIFYPEER, true);
+```
+
+### setCheckSslHost() to setCurlOption()
+
+```diff
+-$client->setCheckSslHost(true);
++$client->setCurlOption(CURLOPT_SSL_VERIFYHOST, 2);
```

Comment:
again, this is a bit to verbose to me. It means the developer is supposed to know about the `CURLOPT_SSL_VERIFYHOST -> 2` detail. Abstracting these kind of details feels like good for the DX (Developer Experience)

Reply:
I'm not sure if having a special method for specific cURL options leads to better DX. I one searches for specific cURL features they will end up on SO or curl.se with instructions to set a specific cURL options. How will one know that there is a specific method instead of setting the cURL option directly with `setCurlOption()`?
Abstracting cURL features will made us care for them. Also where would be the line for which cURL features we should add and which not?
Since this is a Redmine library I would only care about abstract Redmine features (like user impersonation, auth, etc) but leaving the http connection part out of scope of the project.